pkg/auth/server/errorpage/errorpage.go

```diff
@@ -0,0 +1,124 @@
+type ErrorPage struct {
+	render ErrorPageRenderer
+}
+
+// NewErrorPageHandler returns an auth and grant error handler using the given renderer
+func NewErrorPageHandler(renderer ErrorPageRenderer) *ErrorPage {
+	return &ErrorPage{render: renderer}
+}
+
+func (p *ErrorPage) AuthenticationError(err error, w http.ResponseWriter, req *http.Request) (bool, error) {
+	glog.Errorf("AuthenticationError: %v", err)
+	// Only render html error pages for browser-like things
+	if !isBrowser(req) {
+		return false, err
+	}
+
+	errorData := ErrorData{}
+	errorData.ErrorCode = AuthenticationErrorCode(err)
+	errorData.Error = AuthenticationErrorMessage(errorData.ErrorCode)
+
+	p.render.Render(errorData, w, req)
+	return true, nil
+}
+
+func (p *ErrorPage) GrantError(err error, w http.ResponseWriter, req *http.Request) (bool, error) {
+	glog.Errorf("GrantError: %v", err)
+	// Only render html error pages for browser-like things
+	if !isBrowser(req) {
+		return false, err
+	}
+
+	errorData := ErrorData{}
+	errorData.ErrorCode = GrantErrorCode(err)
+	errorData.Error = GrantErrorMessage(errorData.ErrorCode)
+
+	p.render.Render(errorData, w, req)
+	return true, nil
+}
+
+func isBrowser(req *http.Request) bool {
+	switch {
+	// TODO: improve this check and share it with the root redirect
+	case strings.Contains(req.Header.Get("Accept"), "text/html"):
+		return true
+	case strings.Contains(req.UserAgent(), "Mozilla"):
+		return true
+	default:
+		return false
+	}
+}
+
+type ErrorData struct {
+	Error     string
+	ErrorCode string
+}
+
+type ErrorPageRenderer interface {
+	Render(data ErrorData, w http.ResponseWriter, req *http.Request)
+}
+
+type errorPageTemplateRenderer struct {
+	errorPageTemplate *template.Template
+}
+
+// NewErrorPageRenderer creates an error page renderer that takes in an optional custom template to
+// allow branding of the page. Uses the default if templateFile is not set.
+func NewErrorPageTemplateRenderer(templateFile string) (ErrorPageRenderer, error) {
+	r := &errorPageTemplateRenderer{}
+	if len(templateFile) > 0 {
+		customTemplate, err := template.ParseFiles(templateFile)
+		if err != nil {
+			return nil, err
+		}
+		r.errorPageTemplate = customTemplate
+	} else {
+		r.errorPageTemplate = defaultErrorPageTemplate
+	}
+
+	return r, nil
+}
+
+func (r *errorPageTemplateRenderer) Render(data ErrorData, w http.ResponseWriter, req *http.Request) {
+	w.Header().Add("Content-Type", "text/html")
+	w.WriteHeader(http.StatusOK)
+	if err := r.errorPageTemplate.Execute(w, data); err != nil {
+		util.HandleError(fmt.Errorf("unable to render error page template: %v", err))
+	}
+}
+
+func ValidateErrorPageTemplate(templateContent []byte) []error {
+	var allErrs []error
+
+	template, err := template.New("errorPageTemplateTest").Parse(string(templateContent))
+	if err != nil {
+		return append(allErrs, err)
+	}
+
+	var buffer bytes.Buffer
+	err = template.Execute(&buffer, ErrorData{})
```

Comment:
yeah, and didn't want to require them to include either, even in a comment, since they might not want to expose that